### PR TITLE
PMM-4193 Add connectTimeoutMS to MongoDSN.

### PR DESCRIPTION
**Pull request by DexterHD**

### CODE REVIEW

File: models/agent_model.go [hunk 1]
```diff
@@ -172,21 +172,24 @@ func (s *Agent) DSN(service *Service, dialTimeout time.Duration, database string
...
-	case MongoDBExporterType:
-		// TODO return MongoURI for the new driver
-		fallthrough
-
-	case QANMongoDBProfilerAgentType:
+	case QANMongoDBProfilerAgentType, MongoDBExporterType:
 		u := &url.URL{
 			Scheme: "mongodb",
 			Host:   net.JoinHostPort(host, strconv.Itoa(int(port))),
+			Path:   fmt.Sprintf("/%s", database),
```
Comment (AlekSi) [inline models/agent_model.go]: I'm pretty sure we can use `Path: database`.

File: models/agent_model.go [hunk 2]
```diff
@@ -172,21 +172,24 @@ func (s *Agent) DSN(service *Service, dialTimeout time.Duration, database string
...
 			u.User = url.UserPassword(username, password)
 		case username != "":
 			u.User = url.User(username)
 		}
+
+		if dialTimeout > 0 {
+			v := u.Query()
+			v.Set("connectTimeoutMS", strconv.Itoa(int(dialTimeout/time.Millisecond)))
+			u.RawQuery = v.Encode()
+		}
```
Comment (AlekSi) [inline models/agent_model.go]: There is a code below that does the same for PostgreSQL. Is there a reason why we do the same thing in two different ways?

**Comment by DexterHD** [comment]
> Please run tests locally before creating PRs and asking for a review.
Are you sure I'm not doing that?

**Comment by AlekSi** [comment]
I'm pretty sure you did not run tests locally for that PR. https://travis-ci.com/percona/pmm-managed/jobs/209252249 – I have the same failures.

**Comment by DexterHD** [comment]
> I'm pretty sure you did not run tests locally for that PR
No, you are wrong. I'm always running tests locally, not always all tests, but always some of them, related to the task.

Also, there is nothing wrong with don't run it at all as we have CI pipeline for that. It's one of the purposes of CI. To run tests. It does not interfere with code reviews.

**Comment by AlekSi** [comment]
> No, you are wrong. I'm always running tests locally, not always all tests, but always some of them, related to the task.

So help me understand that. In 419aec75b05bb339f6fc4e7ed434afe0034c970d you edited models/agent_model.go to add database name. Related test file models/agent_model_test.go wasn't update. Test failed: https://travis-ci.com/percona/pmm-managed/jobs/209252633 The simplest explanation (even with the benefit of the doubt I can give you) is that you did not in fact run tests.

For the additional context, other build failures I saw yesterday:
* https://travis-ci.com/percona/pmm-agent/jobs/208942525 (vendor is out of sync)
* https://travis-ci.com/percona/mongodb_exporter/jobs/209260502 (test failure)
* https://travis-ci.org/percona/mongodb_exporter/jobs/546717774 (compilation failure)

There is something very wrong with your local setup if it doesn't catch _compilation failures_. Even more concerning that you are unwilling to accept that fact.

> Also, there is nothing wrong with don't run it at all as we have CI pipeline for that. It's one of the purposes of CI. To run tests. It does not interfere with code reviews.

That approach has three associated problems:
1. It creates a huge queue of builds that affect other Percona developers. Making [a fix](https://github.com/percona/mongodb_exporter/pull/145/commits/04d5b3fef59176da4067527be9d7c25f8cb97937) after [fix](https://github.com/percona/mongodb_exporter/pull/145/commits/63437f4265115f1a425b4722bd6fb6e7e4bb6307) could be entirely avoided by a local compilation without blocking everyone else's builds.
2. It makes a feedback loop for you longer. Local compilation and all tests for pmm-managed pass in 27 seconds for the first run, 6 seconds for the second (due to test cache) on my machine. Tests caching is intentionally disabled on Travis CI, and it also much slower, so builds take about 5 minutes each. That's without counting time for waiting in the build queue.
3. Finally, a pull request is just not ready for review if tests do not pass. For example, consider mongodb_exporter at this commit: https://github.com/percona/mongodb_exporter/commit/acf736af74e43a84d229f46062870ec621c48fcf I could say: "ok, cool, great job, approved, just fix a small compilation issue". Then you fixed it by reverting Prometheus client version to the previous one – in PR about "Update prometheus/client library to latest version" nonetheless. What would be a point of the approval then?

In short "there is nothing wrong with don't run it at all as we have CI pipeline for that" is just sloppiness, and it _does_ interfere with code reviews. Don't be sloppy.

Please do not mark pull requests as ready for review unless they are really ready for review (all non-fluky tests pass), or you want to discuss something. Make draft PR if you must, but still please keep in mind points 1 and 2 from above. And hopefully, we will not spend even more time discussing this.
